app/queues/notifications.server.ts

```diff
@@ -0,0 +1,55 @@
+import { Queue, Worker } from 'bullmq';
+
+import { redis } from '~/lib/redis.server';
+import logger from '~/lib/logger.server';
+import send from '~/lib/notifications.server';
+
+import type { Job } from 'bullmq';
+
+export type NotificationsQueueData = {
+  emailAddress: string;
+  subject: string;
+  message: string;
+};
+
+/**
+ * A queue for keeping track of notifications we need to send to users
+ */
+export const notificationsQueue = new Queue<NotificationsQueueData>('notifications', {
+  connection: redis,
+  defaultJobOptions: {
+    attempts: 3,
+    backoff: {
+      type: 'exponential',
+      delay: 60 * 1000 /* 1 minute */,
+    },
+  },
+});
+
+notificationsQueue.on('error', (err) => {
+  logger.warn('notifications queue error', err);
+});
+
+/**
+ * A worker for processing a notification job, sending an email to the user
+ */
+export const notificationsWorker = new Worker<NotificationsQueueData>(
```

Comment:
How do you register this worker into BullMQ? Where is BullMQ initialized?

Reply:
Unless you tell it not to, a `Worker` starts processing as soon as it gets created.  So in this case, the notifications worker will begin waiting for jobs to appear.

Reply:
Okay, so I guess, importing it in the first place causes the global initialization, and the new Worker constructor does also registers it into the global system. Interestint implementation. Thanks for all the explanations!